File: NitroModulesDemo/android/app/src/main/java/com/margelo/nitro/nitromodulesdemo/MathModule.kt
package com.margelo.nitro.nitromodulesdemo

import com.margelo.nitro.core.Promise
import kotlinx.coroutines.delay

/**
 * MathModule implementation in Kotlin.
 * This class inherits from the generated `HybridMathModuleSpec` abstract class.
 */
class MathModule : HybridMathModuleSpec() {
  // Synchronous methods
  override fun add(a: Double, b: Double): Double {
    return a + b
  }

  override fun multiply(a: Double, b: Double): Double {
    return a * b
  }

  // Asynchronous method
  override fun doAsyncWork(): Promise<String> {
    return Promise.async {
      // Simulate some asynchronous work with coroutines
      delay(1500)
      "Async work completed successfully from Kotlin!"
    }
  }

  // Getter method
  override fun getNativeValue(): String {
    return "Value from Kotlin"
  }
}

File: NitroModulesDemo/ios/NitroModulesDemo/MathModule.swift
import Foundation
import NitroModules

/**
 * MathModule implementation in Swift.
 * This class implements the generated `HybridMathModuleSpec` protocol.
 */
class MathModule: HybridMathModuleSpec {
  // Synchronous methods
  func add(a: Double, b: Double) throws -> Double {
    return a + b
  }

  func multiply(a: Double, b: Double) throws -> Double {
    return a * b
  }

  // Asynchronous method
  func doAsyncWork() throws -> Promise<String> {
    return Promise { resolve in
      // Simulate some background work
      DispatchQueue.global().asyncAfter(deadline: .now() + 1.5) {
        resolve("Async work completed successfully from Swift!")
      }
    }
  }

  // Getter method
  func getNativeValue() throws -> String {
    return "Value from Swift"
  }
}
